src/madness/world/madness_exception.h

```diff
@@ -161,6 +160,33 @@ namespace madness {
     } while (0)
 #endif
 
+// the following define/undef are for documentation purposes only.
+/// Check a condition --- even in a release build the condition is always evaluated so it can have side effects
+
+/// Depending on the configuration, one of the following happens if
+/// \c condition is false:
+/// - a \c madness::MadnessException is thrown.
+/// - `check(condition)` is called.
+/// - execution is aborted.
+/// \param[in] condition The condition to be checked.
+#define MADNESS_CHECK(condition)
+#undef MADNESS_CHECK
+
+#ifdef MADNESS_ASSERTIONS_THROW
+#  define MADNESS_CHECK(condition)              \
+    do {                                        \
+        if (!(condition)) {                                             \
+            madness::exception_break(MADNESS_DISPLAY_EXCEPTION_BREAK_MESSAGE); \
+            throw madness::MadnessException("MADNESS CHECKION FAILED: " MADNESS_EXCEPTION_AT( __FILE__, __LINE__ ), \
+                                            #condition,0,__LINE__,__FUNCTION__,__FILE__); \
+        }                                                               \
+    } while (0)
+#else
+#  define MADNESS_CHECK(condition)                      \
+    do {if (!(condition)) { std::abort(); }} while (0)
```

Comment:
probably need to replace `std::abort` to `MPI_Abort` here and everywhere

Reply:
The motivation for std::abort was to force production of a core file.  MPI_ABORT will not routinely do this (e.g., comment from Jeff Squyres who should know https://users.open-mpi.narkive.com/wxLqT6ZJ/ompi-users-abort).   So I am inclined to leave this as is.

In re-reading the code I also noticed that if assertions were configured to do anything other than throw, that checks aborted which is I think not desirable.  So I changed the ifdef logic to make checks always throw unless assertions were configure to abort, in which case both abort.